### DISCUSSION TITLE
Definition of a CNI

### DISCUSSION THREAD

**Discussion by gupta-alok** [lfn-cnti/bestpractices#177] [General]
Based on the discussion in CNF-WG meeting, trying to define the CNI (Container Networking Interface) that can be anchored in [CNCF glossary](https://github.com/cncf/glossary) at some point in time.

So, CNI: Container Networking Interface (CNI) is used to create a generic plugin-based networking solution for containers.

**Comment by taylor** [comment]
CNI (Container Network Interface) is well defined within the K8s community. I suggest we reference what they have.  


From https://github.com/containernetworking/cni
> What is CNI?
> CNI (Container Network Interface), a Cloud Native Computing Foundation project, consists of **a specification and libraries for writing plugins to configure network interfaces in Linux containers**, along with a number of supported plugins. CNI concerns itself only with network connectivity of containers and removing allocated resources when the container is deleted. Because of this focus, CNI has a wide range of support and the specification is simple to implement.

Key part is this:
- **a specification and libraries for writing plugins to configure network interfaces in Linux containers**

Or you could use this
> Container network interface (CNI) plugins are a type of Network plugin that adheres to the appc/CNI specification.
from https://kubernetes.io/docs/reference/glossary/?all=true

---

Refs:
- https://github.com/containernetworking/cni
- https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/
- https://kubernetes.io/docs/reference/glossary/?all=true
- https://github.com/containernetworking/cni
- https://github.com/k8snetworkplumbingwg/sriov-cni

**Reply by jeffsaelens** [reply]
Jason hit us up about helping with the more succinct glossary term for the CNCF repo. He gave us an overview of the targeted audience, and I think @gupta-alok is just kicking things off. 

I do think this is a great way for our team to help contribute to parallel organizations. Coming up with a succinct and easy to understand definition would be super useful I think. 

"CNI: Container Networking Interface (CNI) is used to create a generic plugin-based networking solution for containers." I'd say something more along the lines of "its a specification for creating plugins" and mention that it does some basic plumbing between pods and the kernel.

**Reply by taylor** [reply]
For the high-level, building on what @gupta-alok suggested:

Container Networking Interface (CNI) is a Kubernetes specification and library for creating networking plugins for containers. It takes care of the basic setup connecting the Pods and the host network interfaces. For more detailed information see the [K8s Network Plugin docs](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) or the [CNI github page](https://github.com/containernetworking/cni)

**Reply by jeffsaelens** [reply]
Link to the template of how we should structure it:

https://github.com/cncf/glossary/blob/main/definitions/_TEMPLATE.md

**Reply by taylor** [reply]
---
title: CNI
status: Feedback Appreciated
category: technology
---
## CNI

### What it is
Container Networking Interface (CNI) is a Kubernetes specification and library for creating networking plugins for containers. It takes care of the basic setup connecting the Pods and the host network interfaces. For more detailed information see the [K8s Network Plugin docs](https://kubernetes.io/docs/concepts/extend-kubernetes/compute-storage-net/network-plugins/) or the [CNI github page](https://github.com/containernetworking/cni)

### Problem it Addresses
TBD

### How it Helps
TBD

**Reply by jeffsaelens** [reply]
June 28th WG Call:

Container Networking Interface (CNI) is a container specification and library for creating networking plugins for containers. Kubernetes has adopted the CNI leveraging it for creating network interfaces and allocating IP addresses for pods and other resources. For more detailed information see the K8s Network Plugin docs or the CNI github page

Problem it Addresses
The CNI creates a widely adopted standard for creating networking interfaces within pods. This has alleviated several of the earlier challenges discovered in the container networking space with regards to portability. The CNI provides Layer-3 services, such as connectivity between pods and address management. 

The CNI provides a decoupling of the network specifications of a container from the system or platform hosting them. This decoupling helped facilitate its adoption into the Kubernetes ecosystem which has a strong reliance on "services" to support the dynamic nature of a pod's lifecycle. 

How it Helps
Network developers are able to build against this specification.
